src/mds/Server.cc

```diff
@@ -2893,9 +2893,13 @@ void Server::handle_client_open(MDRequestRef& mdr)
     return;
   }
 
-  // can only open a dir with mode FILE_MODE_PIN, at least for now.
-  if (cur->inode.is_dir())
+  if (!cur->inode.is_file()) {
+    // can only open non-regular inode with mode FILE_MODE_PIN, at least for now.
     cmode = CEPH_FILE_MODE_PIN;
+    // the inode is symlink and client wants to follow it, ignore the O_TRUNC flag.
+    if (cur->inode.is_symlink() && !(flags & O_NOFOLLOW))
+      flags &= ~O_TRUNC;
+  }
```

Comment:
On Fri, May 6, 2016 at 6:43 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/mds/Server.cc
> <https://github.com/ceph/ceph/pull/8778#discussion_r62267169>:
>
> >      cmode = CEPH_FILE_MODE_PIN;
> > +    // the inode is symlink and client wants to follow it, ignore the O_TRUNC flag.
> > +    if (cur->inode.is_symlink() && !(flags & O_NOFOLLOW))
> > +      flags &= ~O_TRUNC;
> > +  }
>
> Is the expected behavior of truncating a symlink documented anywhere? I'm
> not finding it in initial searches, which makes evaluating this behavior a
> little hard.
> (Although it's definitely better than crashing!)
>
see https://github.com/ceph/ceph-client/blob/testing/fs/open.c#L75

But for this case, client actually wants to follow the symlink, truncate
what the symlink points to



—

> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/8778/files/0e4b6f2332bb4822cf324587a94144f1c98e4b97#r62267169>
>